src/glum/_util.py

```diff
@@ -48,13 +53,116 @@ def _align_df_categories(df, dtypes) -> pd.DataFrame:
             changed_dtypes[column] = df[column].cat.set_categories(
                 dtypes[column].categories
             )
+        else:
+            continue
+
+        if cat_missing_method == "convert" and not has_missing_category[column]:
+            unseen_categories = set(df[column].unique()) - set(
+                dtypes[column].categories
+            )
+        else:
+            unseen_categories = set(df[column].dropna().unique()) - set(
+                dtypes[column].categories
+            )
+
+        if unseen_categories:
+            raise ValueError(
+                f"Column {column} contains unseen categories: {unseen_categories}."
+            )
 
     if changed_dtypes:
         df = df.assign(**changed_dtypes)
 
     return df
 
 
+def _add_missing_categories(
+    df,
+    dtypes,
+    feature_names: Sequence[str],
+    categorical_format: str,
+    cat_missing_name: str,
+) -> pd.DataFrame:
+    if not isinstance(df, pd.DataFrame):
+        raise TypeError(f"Expected `pandas.DataFrame'; got {type(df)}.")
+
+    changed_dtypes = {}
+
+    categorical_dtypes = [
+        column
+        for column, dtype in dtypes.items()
+        if pd.api.types.is_categorical_dtype(dtype) and (column in df)
+    ]
+
+    for column in categorical_dtypes:
+        if (
+            categorical_format.format(name=column, category=cat_missing_name)
+            in feature_names
+        ):
+            if cat_missing_name in df[column].cat.categories:
+                raise ValueError(
+                    f"Missing category {cat_missing_name} already exists in {column}."
+                )
+            _logger.info(f"Adding missing category {cat_missing_name} to {column}.")
+            changed_dtypes[column] = df[column].cat.add_categories(cat_missing_name)
```

Comment:
IIRC we add the missing category regardless of the column actually having missing values. So the info message applies even if no observations will belong to the new missing category. I'm not sure if the if block is even necessary. `fillna` simply does not do anything if there are no missings, and it does not really have any performance benefit either, as the colun has to be added to `changed_dtypes` anyways`, right?

Reply:
> `fillna` simply does not do anything if there are no missings

`fillna` will raise an error if the fill value isn't a category of a categorical column, even if the column has no missing values. So we need to add the default category if we call `fillna`. My question was whether we need to add the default category if we don't plan on calling `fillna`.

Reply:
The missing category must be added, even if there are no missings, to accomodate the case in which there are missings in train but not in test (and the reverse case).
I agree that the if block around `fillna` is not necessary, unless `pd.Series.fillna` is much slower than `pd.Series.isnull`.

Reply:
Ah, sorry, I thought the question was about the info message not the `add_categories`. Yes, I think we do need to ad the empty missing category, as this function is used for prediction. If there were missings during training, then we do need to add the corresponding category for the dataset on which we predict, otherwise the coefficient vector and the data matrix are misaligned.

My related thought was that we can probably skip the if condition on line 108, because - conditional on adding the missing category - fillna does not do anything further if there are no missings in the columns.